File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     next: _Node | null
 *     random: _Node | null
 * 
 *     constructor(val?: number, next?: _Node, random?: _Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */


function copyRandomList(head: _Node | null): _Node | null {
    let map = new Map<_Node, _Node>();
    map.set(null, null);

    let start = head;
    while (start) {
        map.set(start, new _Node(start.val));
        start = start.next;
    }

    start = head;
    while (start) {
        let copy = map.get(start);
        copy.next = map.get(start.next);
        copy.random = map.get(start.random);
        start = start.next;
    }

    return map.get(head);
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        mp = {None:None}
        start = head
        while start:
            mp[start] = Node(start.val, None, None)
            start = start.next
        
        start = head
        while start:
            copy = mp.get(start)
            copy.next = mp.get(start.next)
            copy.random = mp.get(start.random)
            start = start.next
        
        return mp.get(head)